File: hackattic/the_sum_of_things/sum_parser.py
import sys

def parse_token(token):
    token = token.strip()

    if not token:
        return 0

    if token.startswith('0x') or token.startswith('0X'):
        return int(token, 16)

    if token.startswith('0o') or token.startswith('0O'):
        return int(token, 8)

    if token.startswith('0b') or token.startswith('0B'):
        return int(token, 2)

    if token.isdigit() or (token.startswith('-') and token[1:].isdigit()):
        return int(token)

    if len(token) == 1:
        return ord(token)

    try:
        return int(token)
    except ValueError:
        return sum(ord(c) for c in token)

def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        tokens = line.split()
        total = sum(parse_token(token) for token in tokens)
        print(total)

if __name__ == "__main__":
    main()

File: hackattic/the_sum_of_things/sum_parser.ex
defmodule SumParser do
  def parse_token(token) do
    token = String.trim(token)

    cond do
      token == "" ->
        0

      String.starts_with?(token, "0x") or String.starts_with?(token, "0X") ->
        hex_digits = String.slice(token, 2..-1//1)
        String.to_integer(hex_digits, 16)

      String.starts_with?(token, "0o") or String.starts_with?(token, "0O") ->
        oct_digits = String.slice(token, 2..-1//1)
        String.to_integer(oct_digits, 8)

      String.starts_with?(token, "0b") or String.starts_with?(token, "0B") ->
        bin_digits = String.slice(token, 2..-1//1)
        String.to_integer(bin_digits, 2)

      Regex.match?(~r/^-?\d+$/, token) ->
        String.to_integer(token)

      String.length(token) == 1 ->
        :binary.first(token)

      true ->
        case Integer.parse(token) do
          {num, ""} -> num
          _ -> token |> to_charlist() |> Enum.sum()
        end
    end
  end

  def process_line(line) do
    line
    |> String.trim()
    |> String.split()
    |> Enum.map(&parse_token/1)
    |> Enum.sum()
  end

  def main do
    IO.stream(:stdio, :line)
    |> Stream.map(&String.trim/1)
    |> Stream.reject(&(&1 == ""))
    |> Stream.map(&process_line/1)
    |> Enum.each(&IO.puts/1)
  end
end

SumParser.main()
